libbpf-tools/ttysnoop.c

```diff
@@ -0,0 +1,299 @@
+
+static const struct argp_option opts[] = {
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ "noclear", 'C', NULL, 0, "Don't clear the screen" },
+	{ "datacount", 'c', "COUNT", 0, "Number of times we check for 'data-size' data (default 16)" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{}
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	switch (key) {
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'C':
+		env.clear_screen = false;
+		break;
+	case 'c':
+		errno = 0;
+		env.count = strtol(arg, NULL, 10);
+		if (errno || env.count <= 0) {
+			warning("Error count: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case ARGP_KEY_ARG: {
+		if (state->arg_num != 0) {
+			warning("Unrecognized positional arguments: %s\n", arg);
+			argp_usage(state);
+		}
+
+		char path[4096] = {};
+		struct stat st;
+
+		if (arg[0] != '/') {
+			strcpy(path, "/dev/pts/");
+			strcat(path, arg);
+		} else {
+			strcpy(path, arg);
+		}
+
+		if (stat(path, &st)) {
+			warning("Failed to stat console file: %s\n", arg);
+			argp_usage(state);
+		}
+		env.pts_inode = st.st_ino;
+		break;
+	}
+	case ARGP_KEY_END:
+		if (env.pts_inode == -1)
+			argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format,
+			   va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = 1;
+}
+
+static int handle_event(void *ctx, void *data, size_t data_sz)
+{
+	const struct event *e = data;
+
+	printf("%s", e->buf);
+	fflush(stdout);
+
+	return 0;
+}
+
+static void handle_lost_events(void *ctx, int cpu, __u64 lost_cnt)
+{
+	warning("Lost %llu event on CPU #%d!\n", lost_cnt, cpu);
+}
+
+static bool fallback_to_compare_kernel_version(void)
+{
+	struct utsname sys_info;
+	int major1, minor1, patch1;
+	int major2, minor2, patch2;
+	const char *version = "5.10.11";
+
+	uname(&sys_info);
+
+	sscanf(sys_info.release, "%d.%d.%d%*s", &major1, &minor1, &patch1);
+	sscanf(version, "%d.%d.%d%*s", &major2, &minor2, &patch2);
+
+	if (major1 < major2)
+		return false;
+	else if (major1 > major2)
+		return true;
+
+	if (minor1 < minor2)
+		return false;
+	else if (minor1 > minor2)
+		return true;
+
+	if (patch1 < patch2)
+		return false;
+	else if (patch1 > patch2)
+		return true;
+
+	return false;
+}
+
+static bool tty_write_is_newly(void)
+{
+	const struct btf_type *type;
+	__s32 id;
+	struct btf *btf;
+
+	btf = btf__load_vmlinux_btf();
+	if (!btf) {
+		warning("No BTF, cannot determine type info: %s", strerror(errno));
+		goto failed;
+	}
+
+	id = btf__find_by_name_kind(btf, "tty_write", BTF_KIND_FUNC);
+	if (id <= 0) {
+		warning("Can't find function tty_write in BTF: %s\n",
+			strerror(-id));
```

Comment:
On arm64 platform, we support load btf from vmlinux, but tty_write is not found in BTF.